File: 프로그래머스/lv1/42862. 체육복/체육복.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(int n, vector<int> lost, vector<int> reserve) {
    std::sort(lost.begin(), lost.end());
    std::sort(reserve.begin(), reserve.end());
    
    for (int i = 1; i <= n; i++) {
        if (std::find(lost.begin(), lost.end(), i) != lost.end() && std::find(reserve.begin(), reserve.end(), i) != reserve.end()) {
            lost.erase(std::remove(lost.begin(), lost.end(), i), lost.end());
            reserve.erase(std::remove(reserve.begin(), reserve.end(), i), reserve.end());
        }
    }

    for (int i = 0; i < reserve.size(); i++) {
        if (reserve[i]-1 >= 1 && std::find(lost.begin(), lost.end(), reserve[i]-1) != lost.end()) {
            lost.erase(std::remove(lost.begin(), lost.end(), reserve[i]-1), lost.end());
            continue;
        }
        if (reserve[i]+1 <= n && std::find(lost.begin(), lost.end(), reserve[i]+1) != lost.end()) {
            lost.erase(std::remove(lost.begin(), lost.end(), reserve[i]+1), lost.end());
        }
        
    }
    return n-lost.size();
}

File: 프로그래머스/1/42862. 체육복/체육복.java
import java.util.*;
class Solution {
    public int solution(int n, int[] lost, int[] reserve) {
        int answer = 0;
        
        Arrays.sort(lost);
        Arrays.sort(reserve);
        
        HashSet<Integer> reserve1 = new HashSet<>();
        HashSet<Integer> lost1 = new HashSet<>();

        for(int num: reserve) {
            reserve1.add(num);
        }
        
        for(int num: lost) {
            lost1.add(num);
        }
        HashSet<Integer> common = new HashSet<>(reserve1);
        common.retainAll(lost1);
        
        reserve1.removeAll(common);
        lost1.removeAll(common);
        
        HashSet<Integer> all = new HashSet<>(reserve1);
        for (int num: lost1) {
            all.add(num);
        }
        
        for (int num: lost1) {
            if (reserve1.contains(num-1)) {
                answer+=2;
                reserve1.remove(num-1);
                continue;
            }
            if (reserve1.contains(num+1)) {
                answer+=2;
                reserve1.remove(num+1);
                continue;
            }
        }
        answer+=reserve1.size();
        answer+=(n-all.size());
        
        return answer;
    }
}